### ISSUE TITLE
sharded unaries are slow and crash on multiple calls

### ISSUE THREAD

**Issue by aguhaTT** [tenstorrent-metal/tt-metal#6937]
Sharded silu and exp have two problems:

1. they are slower than their interleaved counterparts. We expect them to be a few times faster.
2. The program crashes if more than one sharded unary call is placed.

**To Reproduce**
[test_sharded_unary.py.zip](https://github.com/tenstorrent-metal/tt-metal/files/14813535/test_sharded_unary.py.zip)

1. use above test file with the following tests
2. interleaved silu perf is ~10k cycles: `python -m tracy -v -p -r -m pytest test_sharded_unary.py::test_mamba[False-False-5120-silu]`
3. sharded silu perf is 14k cycles, 40% slower: `python -m tracy -v -p -r -m pytest test_sharded_unary.py::test_mamba[True-False-5120-silu]`
4. reproduce silu crash: `python -m tracy -v -p -r -m pytest test_sharded_unary.py::test_mamba[True-True-5120-silu]`
5. interleaved exp perf is 213k cycles: `python -m tracy -v -p -r -m pytest test_sharded_unary.py::test_mamba[False-False-163840-exp]`
6. sharded exp perf is 367k cycles, 72% slower: `python -m tracy -v -p -r -m pytest test_sharded_unary.py::test_mamba[True-False-163840-exp]`
7. reproduce exp crash: `python -m tracy -v -p -r -m pytest test_sharded_unary.py::test_mamba[True-True-163840-exp]`

**Expected behavior**

Sharded >2x faster than interleaved.
No crashes regardless of the number of sharded unary calls in a program. In our case it is 3 (silu, exp, silu).

@kpaigwar @esmalTT

**Comment by jliangTT** [comment]
@shwetankTT , since you have been working on silu, does it make sense to have this fall on your plate?

**Comment by shwetankTT** [comment]
Sure. I will take a look.

**Comment by shwetankTT** [comment]
Hi @aguhaTT 

After running the test suite you provided, I noticed that the interleaved version utilizes 108 cores, while the sharded version uses only 32 cores. This allows the interleaved version to divide the workload into smaller chunks per core, resulting in better performance. I tried using 80 cores (10x8) for both test cases, and now the performance is equivalent. We can increase the number of core further by using padding. Please let me know if my understanding is not correct.

cc @tt-nshanker 
 Please refer to the attached sheet for details.
https://docs.google.com/spreadsheets/d/1-cTHf1b8PO4bMT64NgdRWLBPIGE1xFw0HIx-Er0A-jw/edit?usp=sharing


I am able to reproduce the 2nd issue and working on that.

**Comment by aguhaTT** [comment]
thanks @shwetankTT. I would like to raise a couple of points here. The first one is that we are using WH B0, not GS, so it would be good if you could do your tests on WH. @mywoodstock, in the past, had difficulty, reproducing the same behavior on WH and GS. See my second point below.

**Comment by aguhaTT** [comment]
Let us assume that in this case GS and WH are exhibiting the same behavior. Let us also assume that operations are happening at `hifi4` fidelity, which is the highest and slowest fidelity level. Let us keep aside the comparison of interleaved and sharded for a moment. I realize that comparison was not well thought out. Here are my thoughts why I think `exp`, interleaved or sharded, is too slow. I realize @shwetankTT  may not be responsible for the interleaved version, so it may have to be reassigned. But let us go through the calculations first.

The number of exp ops we need = `32 * 163840 = 2^5 * 2^14 * 10 = 2^19 * 10`
Number of ops that can be done at hifi4 level per cycle per core = `512 = 2^9`
Number of cores used = `32 = 2^5`
Number of cycles = `367000`
Total number of ops that can be done in this time = `2^9 * 2^5 * 367 * 1000 = 2^9 * 2^5 * 2^8 * 2^10 = 2^32` (made an underestimating approximation here)

thus, number of ops involved per `exp` = `2^32 / (2^19 * 10) = 2^13 / 10 = 8192 / 10 = 819.2`

I am not considering data movement, because there should be none for unary ops, either interleaved or sharded.

`819.2` ops (which is an underestimation) to do one `exp` seems way too high. If this is the case, we have to change all our performance projections. 

So, to summarize, there is a performance bug here. I can file a separate issue for it if that helps. Also tagging @uaydonat since this seems to be a higher level problem given that we have to do such a huge `exp`. Basically, it means we have to go back to the drawing board.

**Comment by shwetankTT** [comment]
I have only updated the reader kernel to support the sharded version for the eltwise unary op. There is no change in the compute kernel. I have raised the PR to fix the multiple call crash issue.

We can still track this issue as a performance improvement task.

**Comment by shwetankTT** [comment]
I've updated the sheet with the WH data. Here are a few observations:

- I notice equal performance when the number of cores used for WH is the same.
- In all these runs, the bottleneck appears to be in compute. (compute is taking highest time.)
- The sharded reader kernel is outperforming its interleaved counterpart.

@aguhaTT is there a way we can get the total time/cycle compute kernel was waiting(and busy) for the data?
